File: app/simplex.py

#implmemtacion metodo simplex paso a paso




import numpy as np

def simplex_clasico_paso_a_paso(funcion_objetivo, restricciones, n_vars):
    n_rest = len(restricciones)

    tabla = np.zeros((n_rest + 1, n_vars + n_rest + 1))

    for i, restriccion in enumerate(restricciones):
        tabla[i, :n_vars] = restriccion['coeficientes']
        tabla[i, n_vars + i] = 1  # holgura
        tabla[i, -1] = restriccion['constante']

    tabla[-1, :n_vars] = -np.array(funcion_objetivo)

    pasos = []
    iteracion = 0

    while True:
        iteracion += 1
        paso = {
            "iteracion": iteracion,
            "tabla": tabla.tolist(),
            "pivote_col": None,
            "pivote_fila": None
        }
        col_pivote = np.argmin(tabla[-1, :-1])
        if tabla[-1, col_pivote] >= 0:
            paso["optimo"] = tabla[-1, -1]
            pasos.append(paso)
            break

        ratios = []
        for i in range(n_rest):
            if tabla[i, col_pivote] > 0:
                ratios.append(tabla[i, -1] / tabla[i, col_pivote])
            else:
                ratios.append(np.inf)

        fila_pivote = np.argmin(ratios)
        if ratios[fila_pivote] == np.inf:
            paso["error"] = "Problema no acotado"
            pasos.append(paso)
            break

        paso["pivote_col"] = col_pivote
        paso["pivote_fila"] = fila_pivote
        pasos.append(paso)

        pivote = tabla[fila_pivote, col_pivote]
        tabla[fila_pivote, :] /= pivote
        for i in range(n_rest + 1):
            if i != fila_pivote:
                tabla[i, :] -= tabla[i, col_pivote] * tabla[fila_pivote, :]

    return {"pasos": pasos, "status": "ok"}

File: app/static/js/simplex.js


document.addEventListener("DOMContentLoaded", () => {
    const metodo = document.getElementById("metodo").value.trim()
    if (metodo !== "Simplex") {
        console.log("me fui de simplex")
        return;
    }
    console.log("se ejecuto simplex")
  document.getElementById("formulario").addEventListener("submit", async (e) => {
    e.preventDefault();

    const tipoOperacion = document.getElementById("tipoOperacion").value;
    const funcionObjetivo = document.getElementById("funcionObjetivo").value;
    const numeroVariables = parseInt(document.getElementById("numeroVariables").value);
    const restricciones = [];
    const wrapper = document.getElementById("resWrapper");

    Array.from(wrapper.children).forEach((div, index) => {
      const expr = div.querySelector(`[name^="restriccion_"]`).value;
      const op = div.querySelector(`[name^="operadorRestriccion_"]`).value;
      const val = div.querySelector(`[name^="valorRes_"]`).value;
      restricciones.push({ expr, op, val });
    });

    const payload = { tipoOperacion, funcionObjetivo, numeroVariables, restricciones };
    console.log("payload", payload)
    try {
      const response = await fetch("/resolver_simplex_pasos", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(payload),
      });

      const data = await response.json();

      if (data.status === "ok") {
        mostrarIteraciones(data.pasos);
      } else {
        alert(data.mensaje);
      }

    } catch (err) {
      console.error("Error:", err); //LINEA 44 AQUI <----
    }
  });
});

function mostrarIteraciones(pasos) {
  const contenedor = document.getElementById("resultados");
  contenedor.innerHTML = "";

  pasos.forEach(paso => {
    const div = document.createElement("div");
    div.innerHTML = `<h4>Iteración ${paso.iteracion}</h4>`;
    if (paso.pivote_col !== null) {
      div.innerHTML += `<p>Pivote columna: ${paso.pivote_col + 1}, fila: ${paso.pivote_fila + 1}</p>`;
    }
    const tabla = document.createElement("table");
    tabla.classList.add("table", "table-bordered");
    paso.tabla.forEach(fila => {
      const tr = document.createElement("tr");
      fila.forEach(celda => {
        const td = document.createElement("td");
        td.textContent = parseFloat(celda).toFixed(2);
        tr.appendChild(td);
      });
      tabla.appendChild(tr);
    });
    div.appendChild(tabla);
    contenedor.appendChild(div);
  });
}
